src/main/scala/fs2/kafka/TransactionalMessage.scala

```diff
@@ -0,0 +1,171 @@
+/*
+ * Copyright 2018-2019 OVO Energy Limited
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package fs2.kafka
+
+import cats.{Foldable, Show, Traverse}
+import cats.instances.tuple._
+import cats.syntax.foldable._
+import cats.syntax.show._
+import fs2.kafka.internal.syntax._
+
+/**
+  * [[TransactionalMessage]] represents zero or more `ProducerRecord`s
+  * paired with corresponding `CommittableOffset`s, together with an arbitrary
+  * passthrough value, all of which can be used with [[TransactionalKafkaProducer]].
+  * [[TransactionalMessage]]s can be created using one of the following options.<br>
+  * <br>
+  * - `TransactionalMessage#apply` to produce zero or more records
+  * and then emit a [[ProducerResult]] with the results and
+  * specified passthrough value.<br>
+  * - `TransactionalMessage#one` to produce exactly one record and
+  * then emit a [[ProducerResult]] with the result and specified
+  * passthrough value.<br>
+  * <br>
+  * The [[passthrough]] and [[records]] can be retrieved from an
+  * existing [[TransactionalMessage]] instance.<br>
+  * <br>
+  * For a [[TransactionalMessage]] to be usable by [[TransactionalKafkaProducer]],
+  * it needs a `Traverse[F]` instance. This requirement is
+  * captured in [[TransactionalMessage]] as [[traverse]].
+  */
+sealed abstract class TransactionalMessage[F[_], G[+ _], +K, +V, +P] {
+
+  def records: G[(ProducerRecord[K, V], CommittableOffset[F])]
```

Comment:
Instead of this, I considered making `TransactionalMessage` a subclass of `ProducerMessage` adding a `def committableOffsetBatch: CommittableOffsetBatch[F]` field. Thoughts?